javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
> In the second case the write to value is visible. Then we are finished and can safely return it.

**WRONG**. Let me try to explain this again. You cannot safely return something you have read off the race, you have just constructed the unsafe publication. Even if you have observed the reference to `value`, you are not guaranteed to observe the transitive reads correctly, you have to piggyback on happens-before for this. And to piggy-back on all paths, you *have* to read `supplier` on all paths.

In other words, with "value == null" shortcut, you have just exposed your users to this (I am speculating about the syntax here, sorry, in hurry):

```
  class MyObject {
    int x; 
    MyObject(int x) { this.x = x; }
  }

  static final Lazy<MyObject> lazy = Lazy.with(() -> new MyObject(42));
  ...
  MyObject o = lazy.get(); // discover some other thread had already initalized
  assert (o.x == 42); // intermittently FAILS, because MyObject within Lazy was racily published
```